tracing-appender/src/inner.rs

```diff
@@ -1,80 +1,200 @@
+        let buf_len = buf.len();
+        self.writer.write_all(buf).map(|_| buf_len)
     }
 
     fn flush(&mut self) -> io::Result<()> {
         self.writer.flush()
     }
 }
-
-impl InnerAppender {
-    pub(crate) fn new(
-        log_directory: &Path,
-        log_filename_prefix: &Path,
-        rotation: Rotation,
-        now: DateTime<Utc>,
-    ) -> io::Result<Self> {
+impl InnerAppenderTrait<Roll> for InnerRollingAppender {
+    fn new(roll: Roll, log_directory: &Path, log_filename_prefix: &Path) -> io::Result<Self> {
+        let now = Utc::now();
         let log_directory = log_directory.to_str().unwrap();
         let log_filename_prefix = log_filename_prefix.to_str().unwrap();
 
-        let filename = rotation.join_date(log_filename_prefix, &now);
-        let next_date = rotation.next_date(&now);
+        let filename = roll.join_date(log_filename_prefix, &now);
+        let next_date = roll.next_date(&now);
 
-        Ok(InnerAppender {
+        Ok(InnerRollingAppender {
             log_directory: log_directory.to_string(),
             log_filename_prefix: log_filename_prefix.to_string(),
             writer: create_writer(log_directory, &filename)?,
             next_date,
-            rotation,
+            roll,
         })
     }
 
-    fn write_timestamped(&mut self, buf: &[u8], date: DateTime<Utc>) -> io::Result<usize> {
-        // Even if refresh_writer fails, we still have the original writer. Ignore errors
-        // and proceed with the write.
-        let buf_len = buf.len();
-        self.refresh_writer(date);
-        self.writer.write_all(buf).map(|_| buf_len)
-    }
+    fn refresh_writer(&mut self, _size: &usize) {
+        let now = Utc::now();
 
-    fn refresh_writer(&mut self, now: DateTime<Utc>) {
         if self.should_rollover(now) {
-            let filename = self.rotation.join_date(&self.log_filename_prefix, &now);
+            let filename = self.roll.join_date(&self.log_filename_prefix, &now);
 
-            self.next_date = self.rotation.next_date(&now);
+            self.next_date = self.roll.next_date(&now);
 
             match create_writer(&self.log_directory, &filename) {
                 Ok(writer) => self.writer = writer,
                 Err(err) => eprintln!("Couldn't create writer for logs: {}", err),
             }
         }
     }
+}
 
+impl InnerRollingAppender {
     fn should_rollover(&self, date: DateTime<Utc>) -> bool {
         date >= self.next_date
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct InnerRotatingAppender {
+    log_directory: String,
+    log_filename_prefix: String,
+    writer: BufWriter<File>,
+    last_backup: usize,
+    current_size: usize,
+    rotation: Rotation,
+}
+
+impl io::Write for InnerRotatingAppender {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let buf_len = buf.len();
+        self.writer.write_all(buf).map(|_| buf_len)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.flush()
+    }
```

Comment:
We don't need to implement io::Write again for each appender impl. If you add a default method `record`/`write` then you can just call that from the impl of io::Write for `InnerAppenderWrapper` and not have to duplicate code for every appender.

Reply:
i will still have to create these methods for each appender, I can't create a default method that will access the appenders buffers.